### PR TITLE
Run JavaCC at build to generate TLA+ parser files

### PR DESCRIPTION
**Pull request by ahelwer**
This PR does three things:

- Vendors JavaCC 4.0 to `lib/javacc-4.0.jar` for easy access during grammar development
- Adds a target to the tlatools ant build to generate source files from `tla+.jj`, normalize their line endings, then overwrite the current source files with the newly-generated ones
- Adds a check to the PR CI to ensure `tla+.jj` and the code files are in sync (borrowing method from tree-sitter-tlaplus, which also has generated parser artifacts in source control)

The generated source files are kept in the tree instead of being deleted.

### CODE REVIEW

**Comment by lemmy** [comment]
Is it correct that this breaks our [IDE setup story](https://github.com/tlaplus/tlaplus/tree/master/general/ide)?

**Comment by lemmy** [comment]
One more thing to check is if this causes (additional?) differences between the `tla2tools.jar` when built on Windows, macOS, and Linux, due to the different line endings in the generated files.

**Comment by ahelwer** [comment]
I found an ant build target that normalizes line endings, so that should take care of the build reproducibility issue. I will look into the IDE setup.

**Comment by ahelwer** [comment]
Well I've spent a few hours trying to test that the IDE setup still works and I can't get it to work at baseline. Tried on macOS and ran into #851. Tried on Windows and it gives me a bunch of errors about not being able to find jdk.jfr. So I don't really know what to do here.

**Comment by lemmy** [comment]
I am fine with removing the instructions for setting up the IDE and allowing people to figure out how to configure their development environment on their own.  Even better, let's get rid of the Toolbox.

**Comment by ahelwer** [comment]
Moving this back to draft because in the near future it seems useful to actually see what changes to the output code are made when modifying the grammar file, in a PR.

**Comment by ahelwer** [comment]
@lemmy this PR has made a comeback. I believe it is much more useful and less invasive this time. JavaCC is automatically run to regenerate the grammar files before tla2tools.jar is built, the generated files aren't deleted, and a check is added to the CI to ensure they stay in sync.

**Comment by ahelwer** [comment]
@lemmy @Calvin-L these are some quite simple changes it would be nice to get in!

File: tlatools/org.lamport.tlatools/customBuild.xml [hunk 1]
```diff
@@ -101,7 +99,24 @@
 		<antcall target="dist" inheritall="true" inheritrefs="true" />
 		<antcall target="test-dist-long" inheritall="true" inheritrefs="true" />
 	</target>
-	
+
+	<!-- Runs JavaCC to generate the TLA+ parser files. -->
+	<target name="generate">
+		<tempfile property="javacc-tmpdir" destdir="${java.io.tmpdir}" prefix="javacc_" />
+		<mkdir dir="${javacc-tmpdir}" />
+		<java classpath="lib/javacc-4.0.jar" classname="javacc" fork="true" failonerror="true">
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/customBuild.xml:107]: Have you considered using https://ant.apache.org/manual/Tasks/javacc.html

Comment (ahelwer) [reply tlatools/org.lamport.tlatools/customBuild.xml:107]: I think that's only for later javacc versions

Comment (ahelwer) [reply tlatools/org.lamport.tlatools/customBuild.xml:107]: but I will test it!

Comment (ahelwer) [reply tlatools/org.lamport.tlatools/customBuild.xml:107]: Oh nevermind it wouldn't really work with how we have set up the vendored libs directory.

**Comment by ahelwer** [comment]
Yeah those warnings have been there forever. Maybe I'll get around to figuring out what they mean and fixing them eventually.

**Pull request merged**